File: ios/OpenRUMRNBridge.m
#import "OpenRUMRNBridge.h"
#import <OpenRUM/OpenRUM.h>
#import <OpenCore/OpenCore.h>
#import <React/RCTRootView.h>

@import AlibabaCloudRUM;
@implementation OpenRUMRNBridge

RCT_EXPORT_MODULE(openrumRNBridge);

RCT_EXPORT_METHOD(startWithAppID:(NSString *)appID)
{
    [OpenRUM startWithAppID:appID];
}

RCT_EXPORT_METHOD(setConfigAddress:(NSString *)configAddress)
{
    [OpenRUM setConfigAddress:configAddress];
}

RCT_EXPORT_METHOD(setAppVersion:(NSString *)appVersion)
{
    [OpenRUM setAppVersion:appVersion];
}

RCT_EXPORT_METHOD(setChannelID:(NSString *)channelID)
{
    [OpenRUM setChannelID:channelID];
}

RCT_EXPORT_METHOD(setDeviceID:(NSString *)deviceID)
{
    [OpenRUM setDeviceID:deviceID];
}

RCT_REMAP_METHOD(deviceID,
                 deviceIDWithResolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
{
    NSString *deviceId = [OpenRUM deviceID];
    if (resolve) {
        resolve(deviceId);
    }
}

RCT_REMAP_METHOD(sdkVersion,
                 sdkVersionWithResolver: (RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
{
    NSString *sdkVersion = [OpenRUM SDKVersion];
    if (resolve) {
        resolve(sdkVersion);
    }
}

RCT_EXPORT_METHOD(setUserName:(nullable NSString *)userName)
{
  [AlibabaCloudRUM setUserName:userName];
}

RCT_EXPORT_METHOD(setUserExtraInfo:(nullable NSDictionary<NSString *,id> *)extraInfo)
{
  [AlibabaCloudRUM setUserExtraInfo:extraInfo];
}


RCT_EXPORT_METHOD(setDropFrameTime:(NSInteger)time)
{
    [OpenRUM setDropFrameTime:time];
}

RCT_EXPORT_METHOD(useCustomLaunch:(BOOL)used)
{
    [OpenRUM useCustomLaunch:used];
}

RCT_EXPORT_METHOD(recordCustomLaunchEnd)
{
    [OpenRUM recordCustomLaunchEnd];
}

/// 自定义指标
/// @param metricName 指标名称
/// @param metricValue 指标值（整型）
/// @param param 指标附加信息
RCT_EXPORT_METHOD(setCustomMetric:(NSString *)metricName
                  metricValue:(NSInteger)metricValue
                  param:(nullable NSString *)param)
{
    [OpenRUM setCustomMetricWithName:metricName value:metricValue param:param];
}

/// 自定义日志
/// @param logInfo 日志信息
/// @param param 日志附加信息
RCT_EXPORT_METHOD(setCustomLog:(NSString *)logInfo name:(nullable NSString *)name snapshots:(nullable NSString *)snapshots level:(nullable NSString *)level attributes:(nullable NSDictionary<NSString *,id> *)attributes)
{
  [AlibabaCloudRUM setCustomLog:logInfo name:name snapshots:snapshots level:level info:attributes];
}

RCT_REMAP_METHOD(setCustomEvent,
                 name:(nullable NSString *)name
                 group:(nullable NSString *)group
                 snapshots:(nullable NSString *)snapshots
                 value:(nonnull NSNumber *)value
                 attributes:(nullable NSDictionary *)attributes)
{
  [AlibabaCloudRUM setCustomEvent:name group:group snapshots:snapshots value:value.doubleValue info:attributes];
}

RCT_REMAP_METHOD(setCustomEventStart,
                 setCustomEventStartWithID:(NSString *)eventID
                 name:(nullable NSString *)eventName
                 label:(nullable NSString *)eventLabel
                 param:(nullable NSString *)param
                 info:(nullable NSDictionary *)info)
{
    [OpenRUM setCustomEventStartWithID:eventID
                                   name:eventName
                                  label:eventLabel
                                  param:param
                                  info:info];
}

RCT_REMAP_METHOD(setCustomEventEnd,
                 setCustomEventEndWithID:(NSString *)eventID
                 name:(nullable NSString *)eventName
                 label:(nullable NSString *)eventLabel
                 param:(nullable NSString *)param
                 info:(nullable NSDictionary *)info)
{
    [OpenRUM setCustomEventEndWithID:eventID
                                 name:eventName
                                label:eventLabel
                                param:param
                                info:info];
}

RCT_REMAP_METHOD(setRequestExtraInfo,
                 setRequestExtraInfoWithHeaderKey:(nonnull NSString *)headerKey
                 value:(nonnull NSString *)value
                 info:(nullable NSString *)info)
{
    [OpenRUM setRquestExtraInfoWithHeaderKey:headerKey
                                       value:value
                                        info:info];
}


/// 自定义异常收集
/// @param exceptionType 异常类型
/// @param causedBy 异常原因
/// @param errorDump 异常堆栈
RCT_EXPORT_METHOD(setCustomException:(NSString *)excType
                  causeBy:(nullable NSString *)causeBy
                  errorDump:(nullable NSString *)errorDump)
{
  [AlibabaCloudRUM setCustomException:excType causeBy:causeBy errorDump:errorDump];
}

RCT_REMAP_METHOD(setCustomMethodStart,
                 setCustomMethodStartWithName:(NSString *)methodName
                 param:(nullable NSString *)param)
{
    [OpenRUM setCustomMethodStartWithName:methodName
                                     param:param];
}

RCT_REMAP_METHOD(setCustomMethodEnd,
                 setCustomMethodEndWithName:(NSString *)methodName
                 param:(nullable NSString *)param)
{
    [OpenRUM setCustomMethodEndWithName:methodName param:param];
}

#pragma mark - report

/// 点击事件上报
RCT_EXPORT_METHOD(reportAction:(NSString *)time
                  type:(int)type
                  name:(NSString *)name
                  info:(NSString *)info
                  viewName:(NSString *)viewName
                  loadTime:(NSInteger)loadTime)
{
    long long timeValue = [time longLongValue];
   [OpenRUM reportActionWithTime:timeValue
                             type:type
                             name:name
                             info:info
                         viewName:viewName
                         loadTime:loadTime];
}

///视图事件上报
RCT_EXPORT_METHOD(reportView:(NSString *)time
                  viewId:(NSString *)viewId
                  loadTime:(NSInteger)loadTime
                  model:(int)model
                  name:(NSString *)name
                  methodName:(NSString *)methodName
                  routeId:(NSString *)routeId
                  reactTag:(int)reactTag)
{
    long long timeValue = [time longLongValue];
    [OpenRUM reportViewWithTime:timeValue
                          viewId:viewId
                        loadTime:loadTime
                           model:model
                            name:name
                      methodName:methodName];
  if (model == 1) {//仅进入生成路由Load事件
    [OpenRUM reportRNRouteChangeWithBundleTag:@(reactTag)
                                       routeID:routeId
                                     routeName:name
                                     routeTime:loadTime
                                          time:timeValue];
  }
}

/// 崩溃事件上报
RCT_EXPORT_METHOD(reportCrash:(NSString *)time
                  causeby:(NSString *)causeby
                  type:(NSString *)type
                  dump:(NSString *)dump)
{
    long long timeValue = [time longLongValue];
    
    [OpenRUM reportCrashWithTime:timeValue
                          causeby:causeby
                             type:type
                             dump:dump
                         pageType:0
                        bundleTag:nil
                         anchorID:nil
                        routeName:nil];

}

/// Bundle崩溃事件上报
RCT_EXPORT_METHOD(reportCrashBundleType:(int)bundleType
                  reactTag:(int)reactTag
                  routeId:(NSString *)routeId
                  routeName:(NSString *)routeName
                  time:(NSString *)time
                  causeby:(NSString *)causeby
                  type:(NSString *)type
                  dump:(NSString *)dump)
{
    if (reactTag <= 0) {
        return;
    }
    if (bundleType > 2 || bundleType < 1) {
        return;
    }
    long long timeValue = [time longLongValue];
    [OpenRUM reportCrashWithTime:timeValue
                          causeby:causeby
                             type:type
                             dump:dump
                         pageType:bundleType
                        bundleTag:@(reactTag)
                         anchorID:routeId
                        routeName:routeName];
}

#pragma mark - Private
- (instancetype)init {
  if (self = [super init]) {
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(rootViewDidAppear:) name:RCTContentDidAppearNotification object:nil];
  }
  return self;
}

- (void)rootViewDidAppear:(NSNotification *)noti {
  RCTRootView *rootView = noti.object;
  if (![rootView isKindOfClass:[RCTRootView class]]) {
    return;
  }
  SEL reactTagSelector = NSSelectorFromString(@"reactTag");
  if (![rootView respondsToSelector:reactTagSelector]) {
    return;
  }
  id reactTagNumber = [rootView performSelector:reactTagSelector];
  if (![reactTagNumber isKindOfClass:[NSNumber class]]) {
    return;
  }
  [OpenRUM reportBundleViewAppearTag:reactTagNumber moduleName:rootView.moduleName time:or_cpu_time_us()];
}

- (void)dealloc {
  [[NSNotificationCenter defaultCenter] removeObserver:self];
}



@end

File: android/src/main/java/com/openrum/sdk/bridge/OpenRUMRNBridge.java
package com.openrum.sdk.bridge;

import android.text.TextUtils;
import android.util.Log;

import com.alibabacloud.rum.AlibabaCloudRum;
import com.facebook.react.bridge.ReadableMapKeySetIterator;
import com.openrum.sdk.agent.OpenRum;
import com.openrum.sdk.agent.business.entity.NetworkCustomEventBean;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableArray;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.Promise;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;


public class OpenRUMRNBridge extends ReactContextBaseJavaModule {

    private String configAddress;
    private String appVersion;
    private String channelID;
    private String deviceID;
    private int dropFrameTime;
    private boolean useCustomLaunch;

    public OpenRUMRNBridge(ReactApplicationContext reactContext) {
        super(reactContext);
    }

    @Override
    public String getName() {
        return "openrumRNBridge";
    }

    /**
     * @param time     时间发生时间
     * @param type     操作类型
     * @param name     控件名称
     * @param info     控件信息
     * @param viewName 视图名称
     * @param loadTime 执行耗时
     */
    @ReactMethod
    public void reportAction(String time, int type, String name, String info, String viewName, int loadTime) {
        com.openrum.sdk.agent.engine.external.OpenRUMRNBridge.reportAction(Long.parseLong(time), type, name, info, viewName, loadTime);
    }

    /**
     * @param time       时间发生时间
     * @param viewId     视图id
     * @param loadTime   加载耗时
     * @param model      方式 「1:进入，2:退出」
     * @param name       视图名称
     * @param methodName 方法名称
     */
    @ReactMethod
    public void reportView(String time, String viewId, int loadTime, int model, String name, String methodName, String routeId, int reactTag) {
        com.openrum.sdk.agent.engine.external.OpenRUMRNBridge.reportView(Long.parseLong(time), viewId, loadTime, model, name, methodName, routeId, reactTag);
    }

    /**
     * @param time    时间发生时间
     * @param causeBy 崩溃原因
     * @param type    崩溃类型
     * @param dump    堆栈信息
     */
    @ReactMethod
    public void reportCrash(String time, String causeBy, String type, String dump) {
        com.openrum.sdk.agent.engine.external.OpenRUMRNBridge.reportCrash(Long.parseLong(time), causeBy, type, dump);
    }

    /**
     * @param time    时间发生时间
     * @param causeBy 崩溃原因
     * @param type    崩溃类型
     * @param dump    堆栈信息
     */
    @ReactMethod
    public void reportCrashBundleType(int bundleType, int reactTag, String routeId, String routeName, String time, String causeBy, String type, String dump) {
        Log.d("DEBUGGGGG", "reportCrashBundleType");
        com.openrum.sdk.agent.engine.external.OpenRUMRNBridge.reportCrashBundleType(bundleType, reactTag, routeId, routeName, Long.parseLong(time), causeBy, type, dump);
    }

    @ReactMethod
    public void setCustomLog(String logInfo, String name, String snapshots, String level, ReadableMap attributes) {
        Log.d("DEBUGGGGG", "setCustomLog");
        Map<String, Object> extra = new HashMap<>();
        if (null != attributes) {
            extra = attributes.toHashMap();
        }

        AlibabaCloudRum.setCustomLog(logInfo, name, snapshots, level, extra);
    }

    @ReactMethod
    public void setCustomEvent(String name, String group, String snapshots, double value, ReadableMap info) {
        Log.d("DEBUGGGGG", "setCustomEvent");
        Map<String, Object> extra = new HashMap<>();
        if (null != info) {
            extra = info.toHashMap();
        }
        AlibabaCloudRum.setCustomEvent(name, group, snapshots, value, extra);
        //
        //OpenRum.setCustomEventWithLabel(eventId, eventName, eventLabel, param, extra);
    }

    @ReactMethod
    public void setCustomMetric(String name, double value, String param) {
        try {
            BigDecimal bigDecimal = new BigDecimal(value);
            String str = bigDecimal.toPlainString();
            if (str.contains(".")) {
                str = str.substring(0, str.indexOf("."));
            }
            long time = Long.parseLong(str);
            OpenRum.setCustomMetric(name, time, param);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    /**
     * 冷启动结束调用方法，需要与 withUseCustomLaunch 配合使用，
     * withUseCustomLaunch(true)时，自定义冷启动结束方法生效
     */
    @ReactMethod
    public void recordCustomLaunchEnd() {
        OpenRum.recordCustomLaunchEnd();
    }

    /**
     * userId 数字字母加:_- 按256位限制大小，超过则调用失败
     *
     * @param userName 用户名称
     */
    @ReactMethod
    public void setUserName(String userName) {
        AlibabaCloudRum.setUserName(userName);
    }

    @ReactMethod
    public void setUserExtraInfo(ReadableMap extraInfo) {
        Map<String, Object> extra = new HashMap<>();
        if (null != extraInfo) {
            extra = extraInfo.toHashMap();
        }

        AlibabaCloudRum.setUserExtraInfo(extra);
    }

    @ReactMethod
    public void setCustomMethodStart(String methodName, String param) {
        OpenRum.setCustomMethodStart(methodName, param);
    }

    @ReactMethod
    public void setCustomMethodEnd(String methodName, String param) {
        OpenRum.setCustomMethodEnd(methodName, param);
    }

    /**
     * 自定义异常
     *
     * @param excType   异常类型(必要) limit(0-256字符)
     * @param causeBy   异常原因 limit(0-512字符)
     * @param errorDump 异常信息 limit(0-10000字符)
     */
    @ReactMethod
    public void setCustomException(String excType, String causeBy, String errorDump) {
        AlibabaCloudRum.setCustomException(excType, causeBy, errorDump);
    }

    @ReactMethod
    public void startWithAppID(String appID) {
        OpenRum.withAppID(appID).withConfigAddress(configAddress).withAppVersion(appVersion).withChannelID(channelID).
                withDeviceID(deviceID).withDropFrameTime(dropFrameTime).
                withUseCustomLaunch(useCustomLaunch).start();
    }

    @ReactMethod
    public void setConfigAddress(String configAddress) {
        this.configAddress = configAddress;
    }

    @ReactMethod
    public void setAppVersion(String appVersion) {
        this.appVersion = appVersion;
    }

    @ReactMethod
    public void setChannelID(String channelID) {
        this.channelID = channelID;
    }

    @ReactMethod
    public void setDeviceID(String deviceID) {
        this.deviceID = deviceID;
    }

    @ReactMethod
    public void setDropFrameTime(int time) {
        this.dropFrameTime = time;
    }

    @ReactMethod
    public void useCustomLaunch(boolean useCustomLaunch) {
        this.useCustomLaunch = useCustomLaunch;
    }

    @ReactMethod
    public static void deviceID(Promise promise) {
        String deviceID = OpenRum.getDeviceID();
        if (deviceID != null) {
            promise.resolve(deviceID);
        }
    }

    @ReactMethod
    public static void sdkVersion(Promise promise) {
        String sdkVersion = OpenRum.getSdkVersion();
        if (sdkVersion != null) {
            promise.resolve(sdkVersion);
        }
    }
}
